### PR TITLE
Add GSPO: Sequence-level policy optimization with group advantages

### PR DESCRIPTION
**Pull request by MagellaX**
Overview:
Introduces GSPO (Group Sequence Policy Optimization) as a first-class training mode alongside PPO/GRPO.
Implements sequence-level importance ratio and clipping; supports group-based advantages for multi-sample prompts.

### CODE REVIEW

**Comment by MagellaX** [comment]
Hey @hijkzzz can you merge this?

**Comment by hijkzzz** [comment]
@MagellaX We have already implemented GSPO. What’s the difference with this one?

**Comment by MagellaX** [comment]
> @MagellaX We have already implemented GSPO. What’s the difference with this one?

Wait what? i didn't see it. Are you saying group supervised policy optimization or are u saying that this is GRPO (group_norm/dr_grpo) with token-level PPO loss, cause what my PR brings is sequence‑level importance ratio (mean of token log‑ratios), sequence‑level clipping, and per‑sequence loss reduction, plus a --policy_loss_type gspo flag, this is entirely new sequence-level importance ratio and clipping, with per-sequence reduction

**Comment by MagellaX** [comment]
you can merge it, it's all good!!

File: openrlhf/models/loss.py [hunk 1]
```diff
@@ -112,6 +124,23 @@ def forward(
...
+        else:
+            ratio = (log_probs - old_log_probs).exp()
+            surr1 = ratio * advantages
+            surr2 = ratio.clamp(1 - self.clip_eps_low, 1 + self.clip_eps_high) * advantages
+
+        loss = -torch.min(surr1, surr2)
+
         if self.policy_loss_type == "ppo":
             log_ratio = log_probs - old_log_probs
             ratio = log_ratio.exp()
```
Comment (hijkzzz) [inline openrlhf/models/loss.py:146]: We have implemented GSPO in the next line.

Comment (MagellaX) [reply openrlhf/models/loss.py:146]: Thanks for checking, GSPO needs sequence-level importance ratio and sequence-level clipping. The existing code computes token-level ratio/clipping and only averages after; this PR computes a single ratio per sequence (mean of token log-ratios), clips at the sequence level, and exposes it via --policy_loss_type gspo.
